### ISSUE TITLE
Needless collect suggestion does not compile

### ISSUE THREAD

**Issue by Avanta8** [rust-lang/rust-clippy#8046]
Lint name: `needless_collect`

I tried this code:

```rust
struct A;
impl A {
    fn foo(&mut self) -> i32 {
        0
    }
}

fn f1() {
    let mut a = A {};

    let v = (0..10).map(|_| a.foo()).collect::<Vec<_>>();
    for (i, n) in v.into_iter().enumerate() {
        a.foo();
    }
}
```
As iterators are lazy, I have to collect the iterator to avoid having multiple mutable references to `a`, as `a.foo()` is called within the loop. If this did compile, it could also change the logic of the code if `a.foo` actually did something.

Here is a playgound link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7b7c73582c583be34dabe4deccbdb7ea

I expected to see this happen: The warning would not appear.

Instead, this happened: Clippy wants me to remove the `collect`. However I get compiler error: `cannot borrow `a` as mutable more than once at a time`.

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.46.0-nightly (f455e46ea 2020-06-20)
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0
```

**Comment by mvolfik** [comment]
Previous issue with similar problem: #7975

**Comment by hseg** [comment]
Still present on
```
rustc 1.95.0 (59807616e 2026-04-14) (Arch Linux rust 1:1.95.0-1)
binary: rustc
commit-hash: 59807616e1fa2540724bfbac14d7976d7e4a3860
commit-date: 2026-04-14
host: x86_64-unknown-linux-gnu
release: 1.95.0
LLVM version: 22.1.3
```

triggering on https://github.com/ltratt/pizauth/blob/7dc6962edfcdd7a77f240bd8f5c730a4f6da07fa/src/server/notifier.rs#L68 (the call to `tokenstate_replace` at the end of the loop mutates the underlying `Vec` at the index just inspected)
